src/Raven.Server/Documents/Replication/ReplicationDocumentSender.cs

```diff
@@ -240,6 +244,22 @@ public bool ExecuteReplicationOnce(OutgoingReplicationStatsScope stats, ref Date
 
                             _stats.Storage.RecordInputAttempt();
 
+                            //Here we add missing attachments in the same batch as the document that contains them without modifying the last etag or transaction boundry
+                            if (MissingAttachmentsInLastBatch && item.Type == ReplicationBatchItem.ReplicationItemType.Document &&
+                                (item.Flags & DocumentFlags.HasAttachments) == DocumentFlags.HasAttachments)
+                            {
+                                var type = (item.Flags & DocumentFlags.Revision) == DocumentFlags.Revision ? AttachmentType.Revision: AttachmentType.Document;
+                                foreach (var attachment in _parent._database.DocumentsStorage.AttachmentsStorage.GetAttachmentsForDocument(documentsContext, type, item.Id)
+                                    .Where(a => a.Etag <= firstBatchEtag).Select(a => ReplicationBatchItem.From(a)))
```

Comment:
We can avoid Linq here and filtering it explicitly in the loop - it might be slightly more optimized way

Reply:
this is a rare case and will only invoke if there are missing attachments and if this document has attachments so the performance penalty is neglectable.

Reply:
I don't think we need to do any of that. 
You already _have_ all the attachments you need, in the metadata of the document, no?

Reply:
no

Reply:
why not?

Reply:
The reason we require it here is that we need to send all the attachment's data.
That include the change vector, content-type, etc. These are stored on a per attachment, not at the hash level.

Reply:
However, we don't need to have the `Where` here, we can rely on the fact that attachments are _always_ modified before the document.

Reply:
Therefor, we can remove all the `Linq` in here completely